ros/src/common/libs/state_machine_lib/include/state_machine_lib/state.hpp

```diff
@@ -1,145 +1,152 @@
+    CallbackEntryFunc =
+        [&](const std::string &arg) { /*DEBUG_PRINT("[%s]:%s is not registered\n", state_name_.c_str(), "entry");*/ };
+    CallbackExitFunc =
+        [&](const std::string &arg) { /*DEBUG_PRINT("[%s]:%s is not registered\n", state_name_.c_str(), "exit");*/ };
+#endif
+  }
+  ~State()
   {
-    StateNum = 0;
-    StateTransMask = (uint64_t)STATE_END - 1;
-    StateKind = UNKNOWN_STATE;
   }
 
-public:
-  virtual void update(void)
+  void onUpdate(void)
   {
-    if (StateCallbackUpdateFunc)
-      StateCallbackUpdateFunc();
+    if (child_state_)
+      child_state_->onUpdate();
+
+    if (CallbackUpdateFunc)
+      CallbackUpdateFunc(state_name_);
   }
 
-  virtual void inState(void)
+  void onEntry(void)
   {
-    if (StateCallbackInFunc)
-      StateCallbackInFunc();
+    DEBUG_PRINT("[%s:Entry]\n", state_name_.c_str());
+
+    if (CallbackEntryFunc)
+      CallbackEntryFunc(state_name_);
   }
-  virtual void outState(void)
+  void onExit(void)
   {
-    if (StateCallbackOutFunc)
-      StateCallbackOutFunc();
+    DEBUG_PRINT("[%s:Exit]\n", state_name_.c_str());
+    if (child_state_)
+      child_state_->onExit();
+
+    setChild(nullptr);
+
+    if (CallbackExitFunc)
+      CallbackExitFunc(state_name_);
   }
-  virtual void setCallbackUpdateFunc(std::function<void(void)> _f)
+  void setCallbackUpdate(std::function<void(const std::string&)> _f)
   {
-    StateCallbackUpdateFunc = _f;
+    CallbackUpdateFunc = _f;
   }
-
-  virtual void setCallbackOutFunc(std::function<void(void)> _f)
+  void setCallbackEntry(std::function<void(const std::string&)> _f)
   {
-    StateCallbackOutFunc = _f;
+    CallbackEntryFunc = _f;
   }
 
-  virtual void setCallbackInFunc(std::function<void(void)> _f)
+  void setCallbackExit(std::function<void(const std::string&)> _f)
   {
-    StateCallbackInFunc = _f;
+    CallbackExitFunc = _f;
   }
 
   void showStateName(void)
   {
-    std::cout << StateName << "-";
+    std::cout << state_name_ << "-";
+  }
+  void setParent(std::shared_ptr<State> _parent)
+  {
+    assert(this != _parent.get());
+    parent_state_ = _parent;
   }
 
-  static T* getInstance(void)
+  void setChild(std::shared_ptr<State> _child)
   {
-    static T singleton;
-    return &singleton;
+    assert(this != _child.get());
+    child_state_ = _child;
   }
 
+  std::shared_ptr<State> getParent()
+  {
+    return parent_state_;
+  }
+  std::shared_ptr<State> getChild()
+  {
+    return child_state_;
+  }
   std::string getStateName(void)
   {
-    return std::string(StateName);
+    return std::string(state_name_);
   }
 
-  uint8_t getStateKind(void)
+  void addTransition(const std::string key, const uint64_t val)
   {
-    return StateKind;
+    transition_map_[key] = val;
   }
 
-  uint64_t getStateTransMask(void)
+  uint64_t getTansitionVal(std::string key)
+  {
+    return transition_map_[key];
+  }
+
+  std::map<std::string, uint64_t> getTransitionMap(void)
```

Comment:
Considering that it is a function that has been called multiple times, 
you should output "const &" without making a copy each time.
```suggestion
  const std::map<std::string, uint64_t>& getTransitionMap(void)
```

Reply:
Fixed.